sdks/java/core/src/main/java/org/apache/beam/sdk/util/SdkComponents.java

```diff
@@ -0,0 +1,138 @@
+    this.coderIds = HashBiMap.create();
+  }
+
+  /**
+   * Adds the provided {@link AppliedPTransform} into this {@link SdkComponents}.
+   *
+   * <p>Uses the full name of the transform as the id.
+   */
+  public String getTransformId(AppliedPTransform<?, ?, ?> pTransform) {
+    String existing = transformIds.get(pTransform);
+    if (existing != null) {
+      return existing;
+    }
+    String name = String.format("%s-%s", PTransform.class.getSimpleName(), transformIds.size());
+    transformIds.put(pTransform, name);
+    return name;
+  }
+
+  /**
+   * Puts the provided {@link PCollection} into this {@link SdkComponents} if it is not already
+   * present, returning the id of the provided {@link PCollection} in this {@link SdkComponents}.
+   *
+   * <p>Uses a unique name based on the result of {@link PCollection#getName()}.
+   */
+  public String getPCollectionId(PCollection<?> pCollection) {
+    String existing = pCollectionIds.get(pCollection);
+    if (existing != null) {
+      return existing;
+    }
+    String uniqueName = uniqify(pCollection.getName(), pCollectionIds.values());
+    pCollectionIds.put(pCollection, uniqueName);
+    return uniqueName;
+  }
+
+  /**
+   * Puts the provided {@link WindowingStrategy} into this {@link SdkComponents} if it is not
+   * already present, returning the id of the {@link WindowingStrategy} in this {@link
+   * SdkComponents}.
+   */
+  public String getWindowingStrategyId(WindowingStrategy<?, ?> windowingStrategy) {
+    String existing = windowingStrategyIds.get(windowingStrategy);
+    if (existing != null) {
+      return existing;
+    }
+    String baseName = NameUtils.approximateSimpleName(windowingStrategy);
+    String name = uniqify(baseName, windowingStrategyIds.values());
+    windowingStrategyIds.put(windowingStrategy, name);
+    return name;
+  }
+
+  /**
+   * Puts the provided {@link Coder} into this {@link SdkComponents} if it is not already present,
+   * returning the id of the {@link Coder} in this {@link SdkComponents}.
+   */
+  public String getCoderId(Coder<?> coder) {
+    String existing = coderIds.get(coder);
+    if (existing != null) {
+      return existing;
+    }
+    String baseName = NameUtils.approximateSimpleName(coder);
+    String name = uniqify(baseName, coderIds.values());
+    coderIds.put(coder, name);
+    return name;
+  }
+
+  private String uniqify(String baseName, Set<String> existing) {
+    String name = baseName;
+    int increment = 1;
+    while (existing.contains(name)) {
+      name = baseName + Integer.toString(increment);
+      increment++;
+    }
+    return name;
+  }
+
+  /**
+   * Convert this {@link SdkComponents} into a {@link RunnerApi.Components}, including all of the
+   * contained {@link Coder coders}, {@link WindowingStrategy windowing strategies}, {@link
+   * PCollection PCollections}, and {@link PTransform PTransforms}.
+   */
+  @Experimental
+  public RunnerApi.Components toComponents() {
```

Comment:
Would it better be to just implement a PipelineVisitor that converts from a "java" pipeline to the proto representation, this way you don't need to expose public methods and are effectively walking the graph?